http-clients/src/main/java/org/triplea/http/client/web/socket/GenericWebSocketClient.java

```diff
@@ -0,0 +1,95 @@
+package org.triplea.http.client.web.socket;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.gson.Gson;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.function.Consumer;
+import java.util.logging.Level;
+import lombok.extern.java.Log;
+import org.triplea.java.concurrency.CompletableFutureUtils;
+
+/**
+ * A generic client class for creating websocket connections to a server. This client is based
+ * around a concept of uniform incoming and outgoing messages (eg: an envelope). This class handles
+ * the details of the underlying connection and provides methods to open/close the connection.
+ * Furthermore, this class provides a listener API to be notified of incoming message events, a send
+ * API, and it automatically converts incoming and outgoing messages to JSON string.
+ *
+ * @param <IncomingT> Message type we expect to receive from the server.
+ * @param <OutgoingT> Message type we send to the server.
+ */
+@Log
+public class GenericWebSocketClient<IncomingT, OutgoingT> implements WebSocketConnectionListener {
+  private static final Gson gson = new Gson();
+  private final ExecutorService threadPool = Executors.newFixedThreadPool(4);
```

Comment:
Ok, I'm still not quite getting this, please correct me if I'm missing something.
Why 4 Threads?
The send operation is asynchronous to my knowledge, so using 4 seperate threads shouldn't really speed things up by a noticeable amount, right?

Reply:
It was not really to speed things up, the 4 count is perhaps from earlier decisions and maybe longer applied.

Going into the implementation, we land here to this bottleneck:
```
	private void write( List<ByteBuffer> bufs ) {
		synchronized(synchronizeWriteObject) {
			for( ByteBuffer b : bufs ) {
				write( b );
			}
		}
	}
```
The 4 threads then will certainly not be any faster than 1. Should we then have a threadpool at all? I think yes to avoid blocking on the above synchronize block. It probably makes sense to include the gson operation in the thread as well if we are going to go async sending to our async send queue.

Reply:
Updated to be a threadpool of one and put the gson operation on the new thread: 5e0ca037618220ae3da88ea5f8dc273a52df2055

Added a TODO to revisit the perf gain: c245621542473b56c704d35ca3a6b862a4ae5c17

There are a number of integration tests I'm holding back where it would be easier to verify that, I do not want to integrate those branches with this one and then again with master, and then potentially again if there are yet more changes; there are enough merge conflicts to resolve without re-resolving the same ones multiple times and making them worse with transitory changes if this branches changes further.